### DISCUSSION TITLE
Multiple artists - part one (1h)

### DISCUSSION THREAD

**Discussion by g105b** [SuperHyperInstantFutureTime/TrackShift#63] [Work log]
I'm going to create the concept of `Artist` first by adding a test to the `PRSStatementUploadTest` to test new functionaliy `isMultipleArtist()`. The way I'll implement this is by loading a list of unique Artist objects at the point where the statement is processed. This will help down the line, but for now it'll be really easy to expose this as the `isMultipleArtist()` function.

**Comment by g105b** [comment]
In the `processUsages()` function, each type of Upload can define the column used to identify artists. This column contents can be passed to the `getArtist()` function, which can use a new class called `ArtistIdentifier`, which will help keep things simple further down the line. For now, we'll just treat distinct rows as individual Artists, but in the future we will probably need help identifying artists with similar names across different providers.

**Comment by g105b** [comment]
OK, an Upload now understands whether or not it contains multiple Artists, but before I go any further, I should build the BandcampUpload.

https://github.com/SuperHyperInstantFutureTime/Trackshift/blob/multiple-artists/test/phpunit/Upload/PRSStatementUploadTest.php#L28-L38

**Comment by g105b** [comment]
This might take a while, because I've got to work with the different model of income on the Bandcamp statements - every so often, a sale's revenue is taken by Bandcamp to cover their costs. Biff and I have gone over this in detail so it all makes sense, but I've still got to implement it in a sensible way...

**Reply by g105b** [reply]
What a crazy world we live in. Bandcamp's data is full of invisible characters, probably data that's been copy/pasted somewhere can be detected and traced back to the source. Luckily I was playing with this kind of thing the other day, otherwise these invisible characters would be really confusing.

![image](https://user-images.githubusercontent.com/358014/213255602-035170d0-7e1b-4869-8b72-462e9d96caf1.png)

On upload, the UploadManager will have to strip any control/null characters from the contents to make any sense of it all.

**Reply by g105b** [reply]
We have a `BasecampUpload` being detected and parsed correctly now. There's still a bit of refactoring needed to match the functionality of the PRSStatements.

**Reply by g105b** [reply]
![image](https://user-images.githubusercontent.com/358014/213259850-f1e3749a-e415-4368-8054-af88e83298a0.png)

**Reply by g105b** [reply]
That's all working now. I'm going to write some tests for some cases I've missed and then it's on to the meat of this task... but I'm leaving that for tomorrow.

**Reply by richardbirkin** [reply]
The screenshot / cat / keyboard combo in this pic is AMAZING

**Comment by g105b** [comment]
I need to have a chat with you @richardbirkin about this line: https://github.com/SuperHyperInstantFutureTime/Trackshift/blob/92274f04670df899004400bb8cf4f38edd64ced3/class/Upload/BandcampUpload.php#L23-L27 - you might be able to guess my question from that code.

**Reply by richardbirkin** [reply]
Yeah it's okay to skip it for now as all we want to do is know what each artist has earned 

We may use it in the future to visualise a chronology of bandcamp activity
